benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,229 @@
+from git import Repo
+
+from benchmarks import Benchmark, Result
+
+
+class PersistenceObjectInterface(Protocol):
+    """Interface for constructing the path of a result object."""
+
+    def get_full_file_path(self) -> str:
+        """Construct the path of a result object.
+
+        Returns:
+            The path of the result object.
+        """
+
+    def get_object(self) -> dict:
+        """Construct the object to be persisted.
+
+        Returns:
+            The object to be persisted.
+        """
+
+
+class PersistenceHandlingInterface(Protocol):
+    """Interface for persisting experiment results."""
+
+    def write_object(self, object: PersistenceObjectInterface) -> None:
+        """Store the result of a benchmark.
+
+        Args:
+            object: The object to be persisted.
+        """
+
+
+@define
+class S3PersistenceObjectCreator:
+    """Class for creating the path of a result object."""
+
+    _benchmark: Benchmark = field(validator=instance_of(Benchmark))
+    """The benchmark for which the result is stored."""
+
+    _result: Result = field(validator=instance_of(Result))
+    """The result of the benchmark."""
+
+    _branch: str = field(validator=instance_of(str), init=False)
+    """The branch of the BayBE library from which the workflow was started."""
+
+    _workflow_run_identifier: str = field(validator=instance_of(str), init=False)
+    """The identifier of the workflow run."""
+
+    @_branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        S3_PATH_FORBIDDEN_CHARACTERS = ["/", "\\"]
+        sanitized_branch = ""
+        for char in S3_PATH_FORBIDDEN_CHARACTERS:
+            sanitized_branch = current_branch.replace(char, "-")
+        return sanitized_branch
+
+    @_workflow_run_identifier.default
+    def _default_workflow_id(self) -> str:
+        if "GITHUB_RUN_ID" not in os.environ:
+            raise ValueError("The environment variable GITHUB_RUN_ID is not set.")
+        return os.environ["GITHUB_RUN_ID"]
+
+    def get_full_file_path(self) -> str:
+        """Construct the path of a result object.
+
+        Returns:
+            The path of the result object.
+        """
+        experiment_identifier = self._benchmark.name
+        metadata = self._result.metadata
+        file_usable_date = metadata.start_datetime.strftime("%Y-%m-%d")
+        bucket_path_key = (
+            f"{experiment_identifier}/{self._branch}/{metadata.latest_baybe_tag}/"
+            + f"{file_usable_date}/"
+            + f"{metadata.commit_hash}/{self._workflow_run_identifier}/"
+            + "result.json"
+        )
+        return bucket_path_key
+
+    def get_object(self) -> dict:
+        """Construct the object to be persisted.
+
+        Returns:
+            The object to be persisted.
+        """
+        return self._result.to_dict() | self._benchmark.to_dict()
```

Comment:
What is the rational of storing the entire benchmark and not just it's ID?

Reply:
I think it is a mather of taste and depend on what one would want to have documented and displayed in the dashboard. If I remember correctly, I had a chat with @AVHopp about that. Maybe he can share his opinion. I would want to avoid loading data in a strange way, so the question for me is rather if this data doubling (description, name, and so) for every result disturbs you and should be removed or separated from the varying result.

Reply:
Yes, that is what I mean. There will be redundancy following the current approach. But we might need further components to avoid that or store the benchmark definition once with a dedicated key just for the description.

Reply:
In my opinion, this is fine for now. It semms to do necessary due to the way that responsibilities are handled in the `Benchmark` which will change anyway. So if that is fine for both of you, my vote would be "let's keep it as-is, but add a comment to the code here that this part needs to be changed later on"

Reply:
I think it would be good to implement it directly. Without further API-Abstraction, the storage and its handling is the direct interface for all application. Inconsistency in the way of storing things can be a pain later on.

Reply:
The reason why I did it that way what that things like the metadata may change. So there is not a clear schema to be defined beforehand.

Reply:
A tradeoff might be to only include relevant data from the benchmark (such as the best possible value) and extend the JSON file if necessary. Also, the metadata dict from the result is fully extendable since the dashboard currently add all of its content in a loop. @AdrianSosic @AVHopp Please let me know what you preferer :)

Reply:
For a large scaled system it would properly make sense to combine a filesystem like S3 with a database but I think that is a overkill based on the number of benchmarks and forecasted executions.

Reply:
As already said, my vote is for "Keep-as-is". What about @AdrianSosic resp. is this thread still relevant?

Reply:
Resolving as discussed.